aws/app/lambda/vault_data_integrity_check/vault_data_integrity_check.js

```diff
@@ -0,0 +1,145 @@
+const crypto = require('crypto');
+
+exports.handler = async(event) => {
+
+  try {
+    event.Records
+      .filter(r => r.eventName === "INSERT" || r.eventName === "MODIFY")
+      .filter(r => r.dynamodb.NewImage.NAME_OR_CONF.S.startsWith('NAME#'))
+      .forEach(r => {
+        if (r.eventName === "INSERT") checkInsertEvent(r.dynamodb.NewImage);
+        else checkModifyEvent(r.dynamodb.OldImage, r.dynamodb.NewImage);
+      });
+
+    return {
+      statusCode: "SUCCESS",
+    };
+  }
+  catch (error) {
+    // Error Message will be sent to slack
+    console.error(
+      JSON.stringify({
+        level: "error",
+        severity: 1,
+        msg: "Failed to run Vault data integrity check.",
+        error: error.message,
+      })
+    );
+
+    return {
+      statusCode: "ERROR",
+      error: error.message,
+    };
+  }
+
+};
+
+function checkInsertEvent(data) {
+  let formSubmissionId = "n/a"
+
+  try {
+    formSubmissionId = data?.SubmissionID?.S ?? null;
+    const formSubmissionData = data?.FormSubmission?.S ?? null;
+    const formSubmissionValidHash = data?.FormSubmissionHash?.S ?? null;
+
+    if (!formSubmissionId || !formSubmissionData) {
+      throw new Error("Missing data in record.");
+    }
+
+    if (!formSubmissionValidHash) {
+      console.log(
+        JSON.stringify({
+          level: "warn",
+          msg: `Could not check data integrity for inserted submission (submissionId: ${formSubmissionId}). No valid hash was found in the record. This can happen if the submission was created prior to the introduction of Vault data integrity checks.`,
+        })
+      );
+      return;
+    }
+
+    const formSubmissionHashToValidate = crypto.createHash("md5").update(formSubmissionData).digest("hex");
```

Comment:
There is also the `crypto.verify()` function that can take the hash and data and verifies it within the same function.

Reply:
Looks like the function expects an `ArrayBufferView` for both the data and the hash. It would require some code to convert both strings prior to calling the function.

Reply:
You're right.  That function is more to verify the signature that created the hash.  I was thinking it was similar to the bcrypt.compare function but I was wrong.  Stick with what you have now.